wayland-client/src/event_queue.rs

```diff
@@ -240,6 +244,71 @@ impl<D> EventQueue<D> {
         }
         Ok(dispatched)
     }
+
+    /// Attempt to dispatch events from this queue, registering the current task for wakeup if no
+    /// events are pending.
+    ///
+    /// This method is similar to [`dispatch_pending`](EventQueue::dispatch_pending); it will not
+    /// perform reads on the Wayland socket.  Reads on the socket by other tasks or threads will
+    /// cause the current task to wake up if events are pending on this queue.
+    ///
+    /// ```
+    /// use futures_channel::mpsc::Receiver;
+    /// use futures_util::future::{poll_fn,select};
+    /// use futures_util::stream::StreamExt;
+    /// use wayland_client::EventQueue;
+    ///
+    /// struct Data;
+    ///
+    /// enum AppEvent {
+    ///     SomethingHappened(u32),
+    /// }
+    ///
+    /// impl Data {
+    ///     fn handle(&mut self, event: AppEvent) {
+    ///         // actual event handling goes here
+    ///     }
+    /// }
+    ///
+    /// // An async task that is spawned on an executor in order to handle events that need access
+    /// // to a specific data object.
+    /// async fn run(data: &mut Data, mut wl_queue: EventQueue<Data>, mut app_queue: Receiver<AppEvent>)
+    ///     -> Result<(), Box<dyn std::error::Error>>
+    /// {
+    ///     use futures_util::future::Either;
+    ///     loop {
+    ///         match select(
+    ///             poll_fn(|cx| wl_queue.poll_dispatch_pending(cx, data)),
+    ///             app_queue.next(),
+    ///         ).await {
+    ///             Either::Left((res, _)) => match res? {},
+    ///             Either::Right((Some(event), _)) => {
+    ///                 data.handle(event);
+    ///             }
+    ///             Either::Right((None, _)) => return Ok(()),
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    pub fn poll_dispatch_pending(
+        &mut self,
+        cx: &mut task::Context,
+        data: &mut D,
+    ) -> task::Poll<Result<Infallible, DispatchError>> {
```

Comment:
Is the Ok value supposed to be Infallible or unit?

Unless async runtimes never expect this to return?

Reply:
This particular function will never return an `Ok` variant because you can always keep polling the queue, so `Infallible` is used to denote this.  When using this function, you either need to rely on another future to terminate the task (the end of `app_queue` in the example), or the task will continue running until it gets a `DispatchError` (which might be normal for the "main event queue" of an application)